[ICE]: async closure as _

<!--
[31mICE[0m: Rustc ./a.rs '--edition=2021 -ooutputfile -Zdump-mir-dir=dir' 'thread 'rustc' ($TID) panicked at compiler/rustc_type_ir/src/ty_kind/closure.rs:326:18: 'explicit panic'', 'thread 'rustc' ($TID) panicked at compiler/rustc_type_ir/src/ty_kind/closure.rs:326:18: 'explicit panic''
File: /tmp/im/a.rs
-->

auto-reduced (treereduce-rust):
````rust
//@compile-flags: --edition=2021
fn needs_fn_mut<T>(x: impl FnMut() -> T) {
    needs_fn_mut(async || x as _)
}

````

original:
````rust
fn needs_fn_mut<T>(x: impl FnMut() -> T) {
    needs_fn_mut(async || x as _)
}

fn main() {}

````

Version information
````
rustc 1.97.0-nightly (f53b654a8 2026-04-30)
binary: rustc
commit-hash: f53b654a8882fd5fc036c4ca7a4ff41ce32497a6
commit-date: 2026-04-30
host: x86_64-unknown-linux-gnu
release: 1.97.0-nightly
LLVM version: 22.1.4

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/f53b654a8882fd5fc036c4ca7a4ff41ce32497a6/compiler/rustc_type_ir/src/ty_kind/closure.rs#L320-L332

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc --edition=2021`

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.7aKNqvZVdrhf/rustc_testrunner_tmpdir_reporting.iRHvJdn0oeki/mvce.rs:3:2
  |
3 | }
  |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.7aKNqvZVdrhf/rustc_testrunner_tmpdir_reporting.iRHvJdn0oeki/mvce.rs`

error[E0282]: type annotations needed
 --> /tmp/icemaker_global_tempdir.7aKNqvZVdrhf/rustc_testrunner_tmpdir_reporting.iRHvJdn0oeki/mvce.rs:2:32
  |
2 |     needs_fn_mut(async || x as _)
  |                                ^ cannot infer type


thread 'rustc' (2170575) panicked at /rustc-dev/f53b654a8882fd5fc036c4ca7a4ff41ce32497a6/compiler/rustc_type_ir/src/ty_kind/closure.rs:326:18:
explicit panic
stack backtrace:
   0:     0x7f1e2e2767b9 - <<std[664156f38db6df9a]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[703624273cedab6d]::fmt::Display>::fmt
   1:     0x7f1e2e815f08 - core[703624273cedab6d]::fmt::write
   2:     0x7f1e2e28d1e6 - <std[664156f38db6df9a]::sys::stdio::unix::Stderr as std[664156f38db6df9a]::io::Write>::write_fmt
   3:     0x7f1e2e24c98e - std[664156f38db6df9a]::panicking::default_hook::{closure#0}
   4:     0x7f1e2e269e73 - std[664156f38db6df9a]::panicking::default_hook
   5:     0x7f1e2d260051 - std[664156f38db6df9a]::panicking::update_hook::<alloc[150995461fc6fe17]::boxed::Box<rustc_driver_impl[2526e35b49e1b9f9]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7f1e2e26a152 - std[664156f38db6df9a]::panicking::panic_with_hook
   7:     0x7f1e2e24ca84 - std[664156f38db6df9a]::panicking::panic_handler::{closure#0}
   8:     0x7f1e2e240f89 - std[664156f38db6df9a]::sys::backtrace::__rust_end_short_backtrace::<std[664156f38db6df9a]::panicking::panic_handler::{closure#0}, !>
   9:     0x7f1e2e24e4bd - __rustc[574100d523623398]::rust_begin_unwind
  10:     0x7f1e2aee8c9c - core[703624273cedab6d]::panicking::panic_fmt
  11:     0x7f1e2c12ff02 - core[703624273cedab6d]::panicking::panic
  12:     0x7f1e2e0667e7 - <rustc_type_ir[49665286e8605a96]::ty_kind::closure::CoroutineClosureArgs<rustc_middle[5557c243720a2efc]::ty::context::TyCtxt>>::has_self_borrows
  13:     0x7f1e2e888bec - <rustc_trait_selection[a2580856d63937a4]::traits::select::SelectionContext>::candidate_from_obligation_no_cache
  14:     0x7f1e2f893643 - <rustc_trait_selection[a2580856d63937a4]::traits::select::SelectionContext>::poly_select::{closure#0}
  15:     0x7f1e2ef2526d - <rustc_trait_selection[a2580856d63937a4]::traits::fulfill::FulfillProcessor as rustc_data_structures[215040486e66d119]::obligation_forest::ObligationProcessor>::process_obligation
  16:     0x7f1e2e805a90 - <rustc_data_structures[215040486e66d119]::obligation_forest::ObligationForest<rustc_trait_selection[a2580856d63937a4]::traits::fulfill::PendingPredicateObligation>>::process_obligations::<rustc_trait_selection[a2580856d63937a4]::traits::fulfill::FulfillProcessor>
  17:     0x7f1e2f16075e - rustc_hir_typeck[39142b598aee5fae]::typeck_with_inspect::{closure#0}
  18:     0x7f1e2f15cf34 - rustc_query_impl[53e4749d94278651]::query_impl::typeck_root::invoke_provider_fn::__rust_begin_short_backtrace
  19:     0x7f1e2e9138a6 - rustc_query_impl[53e4749d94278651]::execution::try_execute_query::<rustc_data_structures[215040486e66d119]::vec_cache::VecCache<rustc_span[2d3dd618ebd931a]::def_id::LocalDefId, rustc_middle[5557c243720a2efc]::query::erase::ErasedData<[u8; 8usize]>, rustc_middle[5557c243720a2efc]::dep_graph::graph::DepNodeIndex>, false>
  20:     0x7f1e2e9130c3 - rustc_query_impl[53e4749d94278651]::query_impl::typeck_root::execute_query_non_incr::__rust_end_short_backtrace
  21:     0x7f1e2eb68594 - <rustc_middle[5557c243720a2efc]::ty::context::TyCtxt>::par_hir_body_owners::<rustc_hir_analysis[e68f98e97b2f9de6]::check_crate::{closure#1}>::{closure#0}
  22:     0x7f1e2eb67f66 - rustc_hir_analysis[e68f98e97b2f9de6]::check_crate
  23:     0x7f1e2f7669a1 - rustc_interface[952682f9f3d307b7]::passes::analysis
  24:     0x7f1e2f9a0989 - rustc_query_impl[53e4749d94278651]::execution::try_execute_query::<rustc_middle[5557c243720a2efc]::query::caches::SingleCache<rustc_middle[5557c243720a2efc]::query::erase::ErasedData<[u8; 0usize]>>, false>
  25:     0x7f1e2f9a05e9 - rustc_query_impl[53e4749d94278651]::query_impl::analysis::execute_query_non_incr::__rust_end_short_backtrace
  26:     0x7f1e2f9b7ebc - rustc_interface[952682f9f3d307b7]::interface::run_compiler::<(), rustc_driver_impl[2526e35b49e1b9f9]::run_compiler::{closure#0}>::{closure#1}
  27:     0x7f1e2f99537e - std[664156f38db6df9a]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[952682f9f3d307b7]::util::run_in_thread_with_globals<rustc_interface[952682f9f3d307b7]::util::run_in_thread_pool_with_globals<rustc_interface[952682f9f3d307b7]::interface::run_compiler<(), rustc_driver_impl[2526e35b49e1b9f9]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  28:     0x7f1e2f995a6d - <std[664156f38db6df9a]::thread::lifecycle::spawn_unchecked<rustc_interface[952682f9f3d307b7]::util::run_in_thread_with_globals<rustc_interface[952682f9f3d307b7]::util::run_in_thread_pool_with_globals<rustc_interface[952682f9f3d307b7]::interface::run_compiler<(), rustc_driver_impl[2526e35b49e1b9f9]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[703624273cedab6d]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  29:     0x7f1e2f99686c - <std[664156f38db6df9a]::sys::thread::unix::Thread>::new::thread_start
  30:     0x7f1e2966997a - <unknown>
  31:     0x7f1e296ed2bc - <unknown>
  32:                0x0 - <unknown>

error: the compiler unexpectedly panicked. This is a bug

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.97.0-nightly (f53b654a8 2026-04-30) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z dump-mir-dir=dir

query stack during panic:
#0 [typeck_root] type-checking `needs_fn_mut`
#1 [analysis] running analysis passes on crate `mvce`
end of query stack
error: aborting due to 2 previous errors

Some errors have detailed explanations: E0282, E0601.
For more information about an error, try `rustc --explain E0282`.

```

</p>
</details>

<!--
query stack:
#0 [typeck_root] type-checking `needs_fn_mut`
#1 [analysis] running analysis passes on crate `mvce`

-->

--- a/compiler/rustc_hir_typeck/src/upvar.rs
+++ b/compiler/rustc_hir_typeck/src/upvar.rs
@@ -378,92 +378,108 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         // For coroutine-closures, we additionally must compute the
         // `coroutine_captures_by_ref_ty` type, which is used to generate the by-ref
         // version of the coroutine-closure's output coroutine.
-        if let UpvarArgs::CoroutineClosure(args) = args
-            && !args.references_error()
-        {
-            let closure_env_region: ty::Region<'_> = ty::Region::new_bound(
-                self.tcx,
-                ty::INNERMOST,
-                ty::BoundRegion { var: ty::BoundVar::ZERO, kind: ty::BoundRegionKind::ClosureEnv },
-            );
-
-            let num_args = args
-                .as_coroutine_closure()
-                .coroutine_closure_sig()
-                .skip_binder()
-                .tupled_inputs_ty
-                .tuple_fields()
-                .len();
-            let typeck_results = self.typeck_results.borrow();
-
-            let tupled_upvars_ty_for_borrow = Ty::new_tup_from_iter(
-                self.tcx,
-                ty::analyze_coroutine_closure_captures(
-                    typeck_results.closure_min_captures_flattened(closure_def_id),
-                    typeck_results
-                        .closure_min_captures_flattened(
-                            self.tcx.coroutine_for_closure(closure_def_id).expect_local(),
-                        )
-                        // Skip the captures that are just moving the closure's args
-                        // into the coroutine. These are always by move, and we append
-                        // those later in the `CoroutineClosureSignature` helper functions.
-                        .skip(num_args),
-                    |(_, parent_capture), (_, child_capture)| {
-                        // This is subtle. See documentation on function.
-                        let needs_ref = should_reborrow_from_env_of_parent_coroutine_closure(
-                            parent_capture,
-                            child_capture,
-                        );
-
-                        let upvar_ty = child_capture.place.ty();
-                        let capture = child_capture.info.capture_kind;
-                        // Not all upvars are captured by ref, so use
-                        // `apply_capture_kind_on_capture_ty` to ensure that we
-                        // compute the right captured type.
-                        apply_capture_kind_on_capture_ty(
-                            self.tcx,
-                            upvar_ty,
-                            capture,
-                            if needs_ref {
-                                closure_env_region
-                            } else {
-                                self.tcx.lifetimes.re_erased
-                            },
-                        )
-                    },
-                ),
-            );
-            let coroutine_captures_by_ref_ty = Ty::new_fn_ptr(
-                self.tcx,
-                ty::Binder::bind_with_vars(
-                    self.tcx.mk_fn_sig_safe_rust_abi([], tupled_upvars_ty_for_borrow),
-                    self.tcx.mk_bound_variable_kinds(&[ty::BoundVariableKind::Region(
-                        ty::BoundRegionKind::ClosureEnv,
-                    )]),
-                ),
-            );
-            self.demand_eqtype(
-                span,
-                args.as_coroutine_closure().coroutine_captures_by_ref_ty(),
-                coroutine_captures_by_ref_ty,
-            );
-
-            // Additionally, we can now constrain the coroutine's kind type.
-            //
-            // We only do this if `infer_kind`, because if we have constrained
-            // the kind from closure signature inference, the kind inferred
-            // for the inner coroutine may actually be more restrictive.
-            if infer_kind {
-                let ty::Coroutine(_, coroutine_args) =
-                    *self.typeck_results.borrow().expr_ty(body.value).kind()
-                else {
-                    bug!();
-                };
+        //
+        // If the args already reference an error, computing the by-ref upvar
+        // tuple may itself reach malformed types. We still equate the
+        // `coroutine_captures_by_ref_ty` inference variable to an error type
+        // so downstream consumers (e.g. `has_self_borrows`) can rely on it
+        // being resolved to either an `FnPtr` or `Error` rather than remaining
+        // an unconstrained inference variable.
+        if let UpvarArgs::CoroutineClosure(args) = args {
+            if let Some(guar) = args.error_reported().err() {
                 self.demand_eqtype(
                     span,
-                    coroutine_args.as_coroutine().kind_ty(),
-                    Ty::from_coroutine_closure_kind(self.tcx, closure_kind),
+                    args.as_coroutine_closure().coroutine_captures_by_ref_ty(),
+                    Ty::new_error(self.tcx, guar),
                 );
+            } else {
+                let closure_env_region: ty::Region<'_> = ty::Region::new_bound(
+                    self.tcx,
+                    ty::INNERMOST,
+                    ty::BoundRegion {
+                        var: ty::BoundVar::ZERO,
+                        kind: ty::BoundRegionKind::ClosureEnv,
+                    },
+                );
+
+                let num_args = args
+                    .as_coroutine_closure()
+                    .coroutine_closure_sig()
+                    .skip_binder()
+                    .tupled_inputs_ty
+                    .tuple_fields()
+                    .len();
+                let typeck_results = self.typeck_results.borrow();
+
+                let tupled_upvars_ty_for_borrow = Ty::new_tup_from_iter(
+                    self.tcx,
+                    ty::analyze_coroutine_closure_captures(
+                        typeck_results.closure_min_captures_flattened(closure_def_id),
+                        typeck_results
+                            .closure_min_captures_flattened(
+                                self.tcx.coroutine_for_closure(closure_def_id).expect_local(),
+                            )
+                            // Skip the captures that are just moving the closure's args
+                            // into the coroutine. These are always by move, and we append
+                            // those later in the `CoroutineClosureSignature` helper functions.
+                            .skip(num_args),
+                        |(_, parent_capture), (_, child_capture)| {
+                            // This is subtle. See documentation on function.
+                            let needs_ref = should_reborrow_from_env_of_parent_coroutine_closure(
+                                parent_capture,
+                                child_capture,
+                            );
+
+                            let upvar_ty = child_capture.place.ty();
+                            let capture = child_capture.info.capture_kind;
+                            // Not all upvars are captured by ref, so use
+                            // `apply_capture_kind_on_capture_ty` to ensure that we
+                            // compute the right captured type.
+                            apply_capture_kind_on_capture_ty(
+                                self.tcx,
+                                upvar_ty,
+                                capture,
+                                if needs_ref {
+                                    closure_env_region
+                                } else {
+                                    self.tcx.lifetimes.re_erased
+                                },
+                            )
+                        },
+                    ),
+                );
+                let coroutine_captures_by_ref_ty = Ty::new_fn_ptr(
+                    self.tcx,
+                    ty::Binder::bind_with_vars(
+                        self.tcx.mk_fn_sig_safe_rust_abi([], tupled_upvars_ty_for_borrow),
+                        self.tcx.mk_bound_variable_kinds(&[ty::BoundVariableKind::Region(
+                            ty::BoundRegionKind::ClosureEnv,
+                        )]),
+                    ),
+                );
+                self.demand_eqtype(
+                    span,
+                    args.as_coroutine_closure().coroutine_captures_by_ref_ty(),
+                    coroutine_captures_by_ref_ty,
+                );
+
+                // Additionally, we can now constrain the coroutine's kind type.
+                //
+                // We only do this if `infer_kind`, because if we have constrained
+                // the kind from closure signature inference, the kind inferred
+                // for the inner coroutine may actually be more restrictive.
+                if infer_kind {
+                    let ty::Coroutine(_, coroutine_args) =
+                        *self.typeck_results.borrow().expr_ty(body.value).kind()
+                    else {
+                        bug!();
+                    };
+                    self.demand_eqtype(
+                        span,
+                        coroutine_args.as_coroutine().kind_ty(),
+                        Ty::from_coroutine_closure_kind(self.tcx, closure_kind),
+                    );
+                }
             }
         }
 

--- a/tests/ui/async-await/async-closures/cast-as-infer-self-borrows.rs
+++ b/tests/ui/async-await/async-closures/cast-as-infer-self-borrows.rs
@@ -1,0 +1,10 @@
+//@ edition:2021
+
+// Regression test for #155999.
+
+fn needs_fn_mut<T>(x: impl FnMut() -> T) {
+    needs_fn_mut(async || x as _)
+    //~^ ERROR type annotations needed
+}
+
+fn main() {}

--- a/tests/ui/async-await/async-closures/cast-as-infer-self-borrows.stderr
+++ b/tests/ui/async-await/async-closures/cast-as-infer-self-borrows.stderr
@@ -1,0 +1,9 @@
+error[E0282]: type annotations needed
+  --> $DIR/cast-as-infer-self-borrows.rs:6:32
+   |
+LL |     needs_fn_mut(async || x as _)
+   |                                ^ cannot infer type
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0282`.

